### PR TITLE
Fix tf.math.erf to return ±1 for ±inf instead of nan

### LINKED ISSUE #108483: tf.math.erf returns nan for inf and -inf, inconsistent with PyTorch/SciPy
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

v2.20.0-rc0-4-g72fbba3d20f 2.20.0

### Custom code

Yes

### OS platform and distribution

_No response_

### Mobile device

_No response_

### Python version

_No response_

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

## Description
The TensorFlow function `tf.math.erf(x)` computes the Gauss error function element-wise. However, when the input contains positive or negative infinity (+inf / -inf), the output is nan. This is inconsistent with PyTorch and SciPy, which return 1.0 for +inf and -1.0 for -inf.

## Expected Behavior
For mathematical correctness and consistency with standard libraries, tf.math.erf(+inf) should return 1.0 and tf.math.erf(-inf) should return -1.0.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import numpy as np
import torch
import math
from scipy import special

# TensorFlow behavior
x_tf = tf.constant([np.inf, -np.inf], dtype=tf.float64)
y_tf = tf.math.erf(x_tf)
print("TensorFlow output:", y_tf)
# tf.Tensor([nan nan], shape=(2,), dtype=float64)

# PyTorch behavior
x_torch = torch.tensor(np.array([np.inf, -np.inf]), dtype=torch.float64)
y_torch = torch.erf(x_torch)
print("PyTorch output:", y_torch)
# tensor([1., -1.], dtype=torch.float64)

# Python math library behavior
print(math.erf(float('inf')))   # 1.0
print(math.erf(float('-inf')))  # -1.0

# SciPy behavior
x_np = np.array([np.inf, -np.inf])
y_scipy = special.erf(x_np)
print("SciPy output:", y_scipy)
# [ 1. -1.]
```

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by shahid1330**
## Description
Fixes #108483

This PR fixes `tf.math.erf` to return the mathematically correct values for infinity inputs, making it consistent with PyTorch, SciPy, and Python's math library.

## Problem
`tf.math.erf(±inf)` was returning `nan`, which is incorrect. The mathematical limit of the error function as x approaches ±infinity is ±1.

## Solution
Created a custom erf functor (`scalar_erf_with_inf_op`) that:
- Checks if the input is infinity before computation
- Returns `1.0` for `+inf`
- Returns `-1.0` for `-inf`
- Delegates to Eigen's standard `erf` implementation for all finite values

## Changes
- **tensorflow/core/kernels/cwise_ops.h**: Added `scalar_erf_with_inf_op` functor with infinity handling
- **tensorflow/python/ops/special_math_ops_test.py**: Added `ErfTest` class with comprehensive tests

## Testing
Added unit tests that verify:
1. `erf(+inf) = 1.0` (not nan)
2. `erf(-inf) = -1.0` (not nan)
3. Normal erf behavior for finite values remains unchanged
4. Tests pass for both float32 and float64 dtypes

## Before/After

**Before:**
import tensorflow as tf
import numpy as np
print(tf.math.erf([np.inf, -np.inf]))  # [nan, nan]

**After**
import tensorflow as tf
import numpy as np
print(tf.math.erf([np.inf, -np.inf]))  # [1.0, -1.0]

This makes TensorFlow consistent with:
- **PyTorch**: `torch.erf(torch.inf)` → `1.0`
- **SciPy**: `scipy.special.erf(np.inf)` → `1.0`
- **Python math**: `math.erf(float('inf'))` → `1.0`

### CODE REVIEW

**Comment by mihaimaruseac** [comment]
This seems reasonable to me, but that if might slow down the kernel. So, let's wait for @cantonios to review

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thanks for submitting this pull request! Ensuring `tf.math.erf` returns the correct mathematical limits at infinity is a great improvement and aligns the behavior with PyTorch, SciPy, and Python standard libraries.

However, there are a few issues with the current implementation that need to be addressed before it is ready for merging:

### 1. GPU/TPU Device & Type Compatibility (e.g. `float16` and `bfloat16`)
The line `using std::isinf;` and passing custom types like `Eigen::half` or `bfloat16` to `std::isinf` does not compile or work correctly on GPU/TPU devices because standard library floating-point classification functions are not overloaded/available for these custom types on device backends.
* **Suggestion**: Use `numext::isinf(x)` from the Eigen namespace. It is overloaded for all numeric representations and resolves to the correct platform-specific check for CPU, GPU (CUDA/ROCm), and TPU devices.

### 2. Loss of Vectorization (SIMD)
By creating `scalar_erf_with_inf_op` without a `packetOp` definition and without specializing its `functor_traits`, the Google/TensorFlow cwise runtime falls back to a loop-based non-vectorized execution. This disables vectorization (SIMD) for `tf.math.erf` on CPUs, causing a performance regression for normal finite inputs.
* **Suggestion**: Define `packetOp` using branchless packet operations (e.g., `pisinf`, `pselect`, `pcmp_lt`) to handle infinity values in parallel on CPU vector lanes/GPU threads, and specialize `functor_traits` to specify `PacketAccess = packet_traits<T>::HasErf`.

### Suggested Code Changes

#### `tensorflow/core/kernels/cwise_ops.h`

Please replace the definition of the custom functor and the `erf` struct with the following implementation (defined inside `namespace Eigen::internal`):

```cpp
template <typename T>
struct scalar_erf_with_inf_op {
  EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE T operator()(const T& x) const {
    if (numext::isinf(x)) {
      return x > T(0) ? T(1) : T(-1);
    }
    return scalar_erf_op<T>()(x);
  }
  template <typename Packet>
  EIGEN_DEVICE_FUNC EIGEN_STRONG_INLINE Packet packetOp(const Packet& x) const {
    const Packet zero = pset1<Packet>(T(0));
    const Packet one = pset1<Packet>(T(1));
    const Packet neg_one = pset1<Packet>(T(-1));
    const Packet is_inf = pisinf(x);
    // Determine sign of input elements branchlessly
    const Packet sign = pselect(pcmp_lt(zero, x), one, neg_one);
    const Packet standard = scalar_erf_op<T>().packetOp(x);
    return pselect(is_inf, sign, standard);
  }
};

template <typename T>
struct functor_traits<scalar_erf_with_inf_op<T>> {
  enum {
    Cost = functor_traits<scalar_erf_op<T>>::Cost + 5,
    PacketAccess = packet_traits<T>::HasErf,
  };
};
```

Then define the `erf` struct inside `namespace tensorflow::functor` as:

```cpp
template <typename T>
struct erf : base<T, Eigen::internal::scalar_erf_with_inf_op<T>> {};
```

---

### 3. Unit Tests Coverage (Testing `float16` and `bfloat16`)
We should verify this behavior not just for `float32` and `float64`, but also for `float16` (`half`) and `bfloat16` to ensure complete accuracy across all architectures.
Instead of adding new standalone tests in `special_math_ops_test.py`, it is more appropriate to add the check directly inside the unary operations test suite.

#### `tensorflow/python/kernel_tests/math_ops/cwise_ops_unary_test.py`

Please add this test case to the `UnaryOpTest` class:

```python
  @test_util.run_in_graph_and_eager_modes
  def testErfInfinity(self):
    for dtype in [np.float32, np.float64, np.float16, dtypes.bfloat16.as_numpy_dtype]:
      x = ops.convert_to_tensor([np.inf, -np.inf], dtype=dtype)
      result = math_ops.erf(x)
      expected = np.array([1.0, -1.0], dtype=dtype)
      self.assertAllClose(result, expected)
      self.assertFalse(np.any(np.isnan(self.evaluate(result))))
```

Let me know if you have any questions or need help with implementing this!
